src/models/token/scopes.rs

```diff
@@ -50,3 +51,135 @@ impl FromSql<Text, Pg> for EndpointScope {
         Ok(EndpointScope::try_from(not_none!(bytes))?)
     }
 }
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CrateScope {
+    pattern: String,
+}
+
+impl TryFrom<&str> for CrateScope {
+    type Error = String;
+
+    fn try_from(pattern: &str) -> Result<Self, Self::Error> {
+        match CrateScope::is_valid_pattern(pattern) {
+            true => Ok(CrateScope {
+                pattern: pattern.to_string(),
+            }),
+            false => Err("Invalid crate scope pattern".to_string()),
+        }
+    }
+}
+
+impl TryFrom<String> for CrateScope {
+    type Error = String;
+
+    fn try_from(pattern: String) -> Result<Self, Self::Error> {
+        match CrateScope::is_valid_pattern(&pattern) {
+            true => Ok(CrateScope { pattern }),
+            false => Err("Invalid crate scope pattern".to_string()),
+        }
+    }
+}
+
+impl FromSql<Text, Pg> for CrateScope {
+    fn from_sql(bytes: Option<&[u8]>) -> deserialize::Result<Self> {
+        let value = <String as FromSql<Text, Pg>>::from_sql(bytes)?;
+        Ok(CrateScope::try_from(value)?)
+    }
+}
+
+impl ToSql<Text, Pg> for CrateScope {
+    fn to_sql<W: Write>(&self, out: &mut Output<'_, W, Pg>) -> serialize::Result {
+        <String as ToSql<Text, Pg>>::to_sql(&self.pattern, out)
+    }
+}
+
+impl CrateScope {
+    fn is_valid_pattern(pattern: &str) -> bool {
+        if pattern.is_empty() {
+            return false;
+        }
+
+        if pattern == "*" {
+            return true;
+        }
+
+        let name_without_wildcard = pattern.strip_suffix('*').unwrap_or(pattern);
+        Crate::valid_name(name_without_wildcard)
+    }
+
+    pub fn matches(&self, crate_name: &str) -> bool {
+        let canonicalize = |name: &str| name.replace('-', "_");
```

Comment:
for further discussion: do we want this canonicalization? or do we want the users to be precise?